rust/xaynet-ffi/src/lib.rs

```diff
@@ -1,812 +1,710 @@
+//! Most of the functions have one / or more pointers as parameters.
+//! All functions ensure null-safety for these pointers.
+//! However, when calling this function, you need to ensure that a pointer:
+//! - is properly aligned,
+//! - points to an initialized instance of T where T is the type of data.
+//! The behavior of the function is undefined if the requirements are not met.
+//!
+//! All function of the API are **not** thread-safe.
+//!
+//! # Error handling
+//!
+//! In terms of Error handling, the C-API tries to follow the Posix-style.
+//! Functions return `0` to indicate success and negative values ​​to indicate failure.
+//! Functions that return an opaque pointer (like [`xaynet_ffi_init_mobile_client`])
+//! return a non null pointer to indicate success and a null pointer to indicate failure.
 //!
-//! [`Coordinator`]: ../../coordinator/struct.Coordinator.html
-//! [`Participant`]: ../../participant/struct.Participant.html
 //! [#69173]: https://github.com/rust-lang/rust/issues/69173
 
+#[macro_use]
+extern crate ffi_support;
+
+use ffi_support::FfiStr;
 use std::{
-    ffi::CStr,
-    iter::{IntoIterator, Iterator},
-    mem,
-    os::raw::{c_char, c_int, c_uint, c_ulonglong, c_void},
-    panic,
+    convert::TryFrom,
+    iter::Iterator,
+    os::raw::{c_double, c_int, c_uchar, c_uint, c_void},
     ptr,
+    slice,
 };
 
-use tokio::{
-    runtime::{Builder, Runtime},
-    time::Duration,
+use xaynet_client::mobile_client::{
+    participant::{AggregationConfig, ParticipantSettings},
+    MobileClient,
+};
+use xaynet_core::{
+    certificate::Certificate,
+    crypto::ByteObject,
+    mask::{
+        BoundType,
+        DataType,
+        FromPrimitives,
+        GroupType,
+        IntoPrimitives,
+        MaskConfig,
+        Model,
+        ModelType,
+    },
+    ParticipantSecretKey,
 };
 
-use xaynet_client::{api::HttpApiClient, CachedModel, Client, ClientError, Task};
-use xaynet_core::mask::{FromPrimitives, IntoPrimitives, Model};
-
-#[derive(Clone, Copy, Debug)]
-#[repr(C)]
-/// A fat pointer to a cached model of primitive data type which can be accessed from C.
-///
-/// This is returned from [`new_model()`] and [`get_model()`]. The length `len` will be small enough
-/// such that the respective array fits into memory and the data type `dtype` will be one of the
-/// following:
-/// - `0`: void data type
-/// - `1`: primitive data type [`f32`]
-/// - `2`: primitive data type [`f64`]
-/// - `3`: primitive data type [`i32`]
-/// - `4`: primitive data type [`i64`]
-pub struct PrimitiveModel {
-    /// A raw mutable pointer to an array of primitive values.
-    pub ptr: *mut c_void,
-    /// The length of that array.
-    pub len: c_ulonglong,
-    /// The data type of the array's elements.
-    pub dtype: c_uint,
-}
+/// A Opaque type of MobileClient.
+/// see [FFI-C-OPAQUE](https://anssi-fr.github.io/rust-guide/07_ffi.html#recommendation-a-idffi-c-opaqueaffi-c-opaque)
```

Comment:
It is very nice to have the reference but might be confusing for people reading the doc. Maybe we could turn it into a comment.